### PR TITLE
Add Azure reference implementation

### PR DESCRIPTION
**Pull request by davidmirror-ops**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.terraform*
```
Comment (fiedlerNr9) [inline .gitignore]: We could maybe add *.plan here?
+ out.plan should be deleted

File: environments/azure/flyte-core/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+Example output:
+```bash
+Running Execution on Remote.
+
+[✔] Go to https://flytedeploy01.eastus.cloudapp.azure.com/console/projects/flytesnacks/domains/development/executions/fae18cf6750bd4d64bc7 to see execution in the console.
+```
+5. Go to the console and verify the succesful execution:
+
+![](https://raw.githubusercontent.com/flyteorg/static-resources/main/common/flyte-on-azure-execution.png)  
+
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md:115]: Lets add 6. Tear down Deployment or soomething: `terraform destroy`

File: environments/azure/flyte-core/postgres.tf [hunk 3]
```diff
@@ -0,0 +1,37 @@
+resource "random_password" "postgres" {
+  length           = 20
+  special          = true
+}
+
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/postgres.tf:6]: nit: but this actually happened to me:

```
Error: ---> error converting YAML to JSON: yaml: line 2: found character that cannot start any token userSettings:
│   dbPassword: @=2o4*DZFsCy#+eJ{5RT
│   dbUser: flyte
│   dbName: flyte
│   dbHost: jan-deployment-flyte.postgres.database.azure.com
│   storageContainerName: flytetf
│   storageAccountName: jandeploymentflytetf
│   dnsLabel: jan.eastus.cloudapp.azure.com
│   backendWIClientId: e3f726c4-d243-41ff-b33f-7770376635fa
│   tasksWIClientId: 6d9e64db-63bb-4cfa-a701-44b6a3be3553
```
Maybe we just get rid of special chars

File: environments/azure/flyte-core/flyte-core.tf [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    storage_account_container_name = azurerm_storage_container.flyte.name
+    storage_account_name           = azurerm_storage_account.flyte.name
+    dns_label                      = "${local.flyte_domain_label}.${local.location}.cloudapp.azure.com"
+    backend_wi_client_id           = azuread_application.flyte_backend.client_id
+    tasks_wi_client_id             = azuread_application.flyte_tasks.client_id
+    }
+    )
+  ]
+  depends_on = [azurerm_postgresql_flexible_server_firewall_rule.all, azurerm_postgresql_flexible_server_database.flyte, azurerm_role_assignment.backend_role_assignment]
+  provisioner "local-exec" {
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/flyte-core.tf]: getting the kubeconfig should happen after the AKS creation right? 

So this local-exec needs to go aks.tf

File: environments/azure/flyte-core/ingress-tls.tf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+   set{
+    name = "controller.service.annotations.service\\.beta\\.kubernetes\\.io/azure-load-balancer-health-probe-request-path"
+    value = "/healthz"
+   }
+    provisioner "local-exec" {
+      command = "./scripts/connect_flyte.sh ${azurerm_resource_group.flyte.name} ${azurerm_kubernetes_cluster.flyte.name} ${local.flyte_domain_label}"
+ }
+ }
+
+resource kubernetes_namespace "cert_manager_ns"{
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/ingress-tls.tf]: Is this the only resource coming from the kubernetes provider? If yes, we could also build the namespace with the kubectl provider and ending up with -1 provider

File: environments/azure/flyte-core/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+#Uncomment only if you want to test CLI commands and the certificate is not generated yet.
+# You can confirm the cert by either going to the UI (a valid certificate should be used) or
+#from your terminal: kubectl get challenges.acme.cert-manager.io -n flyte (there should not be any pending challenge). With this flag enabled, SSL is still used but the client doesn't verify the certificate chain.
+
+  #insecureSkipVerify: true 
+```
+> NOTE: this configuration step is only needed for CLI access (`flytectl` or `pyflyte`), not for the UI.
+
+
+3. Save the following "hello world" workflow definition:
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md:87]: This command does not work for me. I think some kind of string is not ending properly

File: environments/azure/flyte-core/providers.tf [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  load_config_file       = false
+}
+provider "azurerm" {
+  features {}
+
+  subscription_id = local.subscription_id
+  tenant_id       = local.tenant_id
+  storage_use_azuread = true
+  use_aks_workload_identity = true
+ # use_cli                   = false
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/providers.tf]: Why do we need these 2 attributes here?

```
storage_use_azuread = true
use_aks_workload_identity = true
```

File: environments/azure/flyte-core/values-aks.yaml [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+  custom:
+    container: "{{ .Values.userSettings.storageContainerName }}"
+    enable-multicontainer: true
+    connection: {}
+    type: stow
+    stow:
+      kind: azure
+      config:
+        account: "{{ .Values.userSettings.storageAccountName }}"
+      #key: "{{ .Values.userSettings.storageAccountKey }}" #Hopefully we'll switch this to proper IAM
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/values-aks.yaml]: > Hopefully we'll switch this to proper IAM

It is proper IAM now right?

File: environments/azure/flyte-core/iam.tf [hunk 9]
```diff
@@ -0,0 +1,89 @@
+
+data "azuread_client_config" "current" {}
+
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/iam.tf]: On the topic of using workload identities with AKS. I think there are two possibilities to enable this:
Service Principals(your approach) vs Managed Identities (I like these more).

[What the difference?](https://devblogs.microsoft.com/devops/demystifying-service-principals-managed-identities/)

On this one we can chat in private but i think going with Managed Identities is more save (no credentials) and requires less terraform code to setup.

**Review by fiedlerNr9 — CHANGES_REQUESTED** [review]
This is amazing stuff! The Ingress & TLS automation is something i want to put on my wall ;) Left some comments

File: environments/azure/flyte-core/README.md (hunk 6 above)
Comment (davidmirror-ops) [reply environments/azure/flyte-core/README.md:87]: Right! There was an extra `<`

File: environments/azure/flyte-core/iam.tf (hunk 9 above)
Comment (davidmirror-ops) [reply environments/azure/flyte-core/iam.tf]: I originally went with the Managed Identities approach but the issue was that role assignment was only possible for a Service Principal and not a MI. The Flyte users who have Flyte running on AZ are using SPs. But I agree MI would be better. I'll give it a second try and will update here accordingly. Thanks!

**Review by fiedlerNr9 — APPROVED** [review]
This is awesome stuff!

**Pull request merged**